atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -1330,33 +1349,95 @@ private static Integer getMaxRowsFromBatchHint(@Nullable Integer batchHint) {
         T run(DbReadTable readTable, DbWriteTable writeTable);
     }
 
-    static class ResultWithToken {
-        SetMultimap<Cell, Long> results;
-        byte[] currentRow = null;
-        boolean finishedRange = false;
+    static final class ResultWithToken {
+        private byte[] currentRow = null;
+        private byte[] currentCol = null;
+        private Long currentTimestamp = null;
+        private PeekingIterator<AgnosticLightResultRow> iterator;
+
+        final SetMultimap<Cell, Long> entries;
+        boolean mayHaveMoreResults = false;
+        Token token = Token.INITIAL;
+
+        private ResultWithToken(ClosableIterator<AgnosticLightResultRow> iterator) {
+            entries = HashMultimap.create();
+            this.iterator = Iterators.peekingIterator(iterator);
+        }
+
+        static ResultWithToken create(ClosableIterator<AgnosticLightResultRow> iterator) {
+            return new ResultWithToken(iterator);
+        }
+
+        /**
+         * @param oldToken token from previous page, specifying if we have already processed some entries from the
+         * current row and should therefore skip them. If oldToken.shouldForward() is true, we iterate until the end or
+         * the first result that is either:
+         *  1. In another row
+         *  2. In a greater column
+         *  3. In the same column, with higher or equal timestamp (to repeat the last timestamp for sweep)
+         */
+        void moveForward(Token oldToken) {
+            boolean forwarding = oldToken.shouldForward();
```

Comment:
Perhaps rename `forwarding` to `movingForward` and `shouldForward` to `shouldMoveForward`?